File: packages/ts/cairn-p2p/src/pairing/adapter.ts
/**
 * Custom pairing adapter interface for domain-specific pairing flows.
 *
 * Applications implement this interface for custom transports:
 * NFC tap, Bluetooth LE OOB, email-based verification, hardware token, etc.
 */
export interface PairingMechanismAdapter {
  /** Human-readable name of this mechanism (e.g., "nfc", "bluetooth-le"). */
  name: string;

  /** Create the pairing payload in the application's chosen format/transport. */
  generatePayload(data: Uint8Array): Promise<Uint8Array>;

  /**
   * Parse and validate a received pairing payload from the custom transport.
   * Returns the PAKE credential and optional connection hints.
   */
  consumePayload(data: Uint8Array): Promise<{
    pakeCredential: Uint8Array;
    hints?: string[];
  }>;

  /** Derive the SPAKE2 password bytes from the custom payload data. */
  derivePakeInput(data: Uint8Array): Promise<Uint8Array>;
}

/**
 * Wrapper that bridges a PairingMechanismAdapter into the pairing system.
 */
export class CustomPairingMechanism {
  constructor(private readonly adapter: PairingMechanismAdapter) {}

  get name(): string {
    return this.adapter.name;
  }

  async generatePayload(data: Uint8Array): Promise<Uint8Array> {
    return this.adapter.generatePayload(data);
  }

  async consumePayload(data: Uint8Array): Promise<{
    pakeCredential: Uint8Array;
    hints?: string[];
  }> {
    return this.adapter.consumePayload(data);
  }

  async derivePakeInput(data: Uint8Array): Promise<Uint8Array> {
    return this.adapter.derivePakeInput(data);
  }
}

File: packages/go/cairn-p2p/pairing/adapter.go
package pairing

import (
	"github.com/moukrea/cairn/packages/go/cairn-p2p/crypto"
)

// PairingPayload is a generic pairing payload used by adapters.
type PairingPayload struct {
	PeerID   [34]byte
	Nonce    [16]byte
	PakeCred []byte
	Hints    []string
}

// PairingAdapter allows applications to provide custom pairing flows
// (e.g., NFC, Bluetooth LE, email-based, hardware token).
type PairingAdapter interface {
	// GeneratePayload creates a pairing payload for the given identity.
	GeneratePayload(identity *crypto.IdentityKeypair) ([]byte, error)

	// ConsumePayload parses raw bytes into a PairingPayload.
	ConsumePayload(data []byte) (*PairingPayload, error)

	// IntegrateKeyExchange integrates a payload into the key exchange,
	// returning the PAKE credential bytes.
	IntegrateKeyExchange(payload *PairingPayload) ([]byte, error)
}
